src/test/java/HeapTest.java

```diff
@@ -0,0 +1,433 @@
+                }
+
+                @Test
+                void isEmpty_singleton_returnsFalse() {
+                    assertThat(classUnderTest.isEmpty()).isFalse();
+                }
+
+                @Test
+                void toString_singleton_returnsString() {
+                    assertThat(classUnderTest.toString()).isEqualTo("[10]");
+                }
+
+                @Nested
+                @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+                class WhenMany {
+
+                    @BeforeEach
+                    void addMany() {
+                        classUnderTest.add(20);
+                        classUnderTest.add(20);
+                        classUnderTest.add(10);
+                        classUnderTest.add(5);
+                        classUnderTest.add(15);
+                        preState.add(20);
+                        preState.add(20);
+                        preState.add(10);
+                        preState.add(5);
+                        preState.add(15);
+                    }
+
+                    @Test
+                    void add_many_returnsTrue() {
+                        assertThat(classUnderTest.add(99)).isTrue();
+                    }
+
+                    @Test
+                    void remove_many_returnsElement() {
+                        assertThat(classUnderTest.remove()).isEqualTo(5);
+                    }
+
+                    @Test
+                    void remove_many_change() {
+                        Heap<Integer> expected = new Heap<>();
+                        expected.add(10);
+                        expected.add(10);
+                        expected.add(15);
+                        expected.add(20);
+                        expected.add(20);
+                        classUnderTest.remove();
+                        assertThat(classUnderTest).isEqualTo(expected);
+                    }
+
+                    @Test
+                    void peek_many_returnsElement() {
+                        assertThat(classUnderTest.peek()).isEqualTo(5);
+                    }
+
+                    @Test
+                    void peek_many_unchanged() {
+                        classUnderTest.peek();
+                        assertThat(classUnderTest).isEqualTo(preState);
+                    }
+
+                    @Test
+                    void size_many_returnsSize() {
+                        assertThat(classUnderTest.size()).isEqualTo(6);
+                    }
+
+                    @Test
+                    void isEmpty_many_returnsFalse() {
+                        assertThat(classUnderTest.isEmpty()).isFalse();
+                    }
+
+                    @Test
+                    void toString_many_returnsString() {
+                        assertThat(classUnderTest.toString()).isEqualTo("[5, 10, 15, 20, 10, 20]");
+                    }
+                }
+            }
+        }
+    }
+
+    @Nested
+    class ComparatorConstructor {
```

Comment:
More simple imo is to simply remove the default constructor, and always accept a comparator. This also removes the whole "_might_ throw a `ClassCastException` sometimes" thing, pushing the error to the user for providing an invalid comparator.

Reply:
Brilliant.